content/blog/2025-12-18-extending-sql.md

```diff
@@ -0,0 +1,379 @@
+  .build();
+```
+
+Once installed, if your `CREATE EXTERNAL CATALOG` statement exposes tables with MySQL types, DataFusion can interpret them correctly.
+
+---
+
+## 4) Extending the FROM clause: `RelationPlanner`
+
+Some extensions change what a _relation_ means, not just expressions or types. `RelationPlanner` intercepts FROM-clause constructs while SQL is being converted into a `LogicalPlan`.
+
+`RelationPlanner` originally came out of trying to build `MATCH_RECOGNIZE` support in DataFusion as a Datadog hackathon project. `MATCH_RECOGNIZE` is a complex SQL feature for detecting patterns in sequences of rows, and it made sense to prototype as an extension first. At the time, DataFusion had no extension point at the right stage of SQL-to-rel planning to intercept and reinterpret relations.
+
+[@theirix]'s `TABLESAMPLE` work ([#13563], [#17633]) demonstrated exactly where the gap was: the extension only worked when `TABLESAMPLE` appeared at the query root and any `TABLESAMPLE` inside a CTE or JOIN would error. That limitation motivated [#17843], which introduced `RelationPlanner` to intercept relations at any nesting level. The same hook now supports `PIVOT`, `UNPIVOT`, `TABLESAMPLE`, and can translate dialect-specific FROM-clause syntax (for example, bridging Trino constructs into DataFusion plans).
+
+This is how Datadog approaches compatibility work: build features in real systems first, then upstream the building blocks. A full `MATCH_RECOGNIZE` extension is now in progress, built on top of `RelationPlanner`, with the [`match_recognize.rs`][match_recognize.rs] example as a starting point.
+
+Once you have `RelationPlanner`, there are two main approaches to implementing your extension.
+
+### Strategy A: rewrite to existing operators (PIVOT / UNPIVOT)
+
+If you can translate your syntax into relational algebra that DataFusion already supports, you can implement the feature with **no custom physical operator**.
+
+`PIVOT` rotates rows into columns, and `UNPIVOT` does the reverse. Neither requires new execution logic: `PIVOT` is just `GROUP BY` with `CASE` expressions, and `UNPIVOT` is a `UNION ALL` of each column. The planner rewrites them accordingly:
+
+```rust
+match relation {
+  TableFactor::Pivot { .. } => /* rewrite to GROUP BY + CASE */,
+  TableFactor::Unpivot { .. } => /* rewrite to UNION ALL */,
+  other => Original(other),
+}
+```
+
+Because the output is a standard `LogicalPlan`, DataFusion's usual optimization and physical planning apply automatically.
+
+**Full working example:** [`pivot_unpivot.rs`][pivot_unpivot.rs]
+
+### Strategy B: custom logical + physical (TABLESAMPLE)
+
+Sometimes rewriting is not sufficient. `TABLESAMPLE` returns a random subset of rows from a tableand is useful for approximations or debugging on large datasets. Because it requires runtime randomness, you cannot express it as a rewrite to existing operators. Instead, you need a custom logical node and physical operator to execute it.
```

Comment:
```suggestion
Sometimes rewriting is not sufficient. `TABLESAMPLE` returns a random subset of rows from a table and is useful for approximations or debugging on large datasets. Because it requires runtime randomness, you cannot express it as a rewrite to existing operators. Instead, you need a custom logical node and physical operator to execute it.
```